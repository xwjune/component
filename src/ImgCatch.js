/**
 * 图片异常捕获
 *
 * @param {string} src - 图片资源
 * @param {string} [blanksrc] - 空白页图片资源
 * @param {function} [onError] - 回调
 * @param {*} - img其他属性
 *
 * @author 小巷 <xwjune@163.com>
 * @date 2019/01/15
 */
import React from 'react';
import PropTypes from 'prop-types';
import imgBlank from './assets/blank_img.svg';

export default class ImgCatch extends React.PureComponent {
  static propTypes = {
    src: PropTypes.string,
    blanksrc: PropTypes.string,
    onError: PropTypes.func,
  };

  static defaultProps = {
    blanksrc: imgBlank,
  };

  state = {
    imgErr: false,
  };

  onError = () => {
    if (!this.state.imgErr) {
      this.setState({
        imgErr: true,
      });
      if (this.props.onError) {
        this.props.onError();
      }
    }
  };

  render() {
    return (
      <img
        alt="图片"
        {...this.props}
        src={this.state.imgErr ? this.props.blanksrc : this.props.src}
        onError={this.onError}
      />
    );
  }
}
