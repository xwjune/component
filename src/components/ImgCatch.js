/**
 * 图片异常捕获
 *
 * @param {string} src - 图片资源
 * @param {string} [blanksrc] - 空白页图片资源
 * @param {function} [onError] - 回调
 * @param {*} - img其他属性
 *
 * @date 2019/01/15
 */
import React from 'react';
import PropTypes from 'prop-types';

export default class ImgCatch extends React.PureComponent {
  static propTypes = {
    src: PropTypes.string,
    blanksrc: PropTypes.string.isRequired,
    onError: PropTypes.func,
  };

  state = {
    imgErr: false,
  };

  onError = () => {
    if (!this.state.imgErr) {
      this.setState({
        imgErr: true,
      });
      if (this.props.onError) {
        this.props.onError();
      }
    }
  };

  render() {
    return (
      <img
        alt="图片"
        {...this.props}
        src={this.state.imgErr ? this.props.blanksrc : this.props.src}
        onError={this.onError}
      />
    );
  }
}
