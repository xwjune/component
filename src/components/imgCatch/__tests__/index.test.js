import React from 'react';
import { shallow } from 'enzyme';
import ImgCatch from '../index';

describe('ImgCatch', () => {
  test('renders correctly', () => {
    const wrapper = shallow(
      <ImgCatch
        width="100"
        height="100"
        src="pic.png"
        blanksrc="http://pic/blanksrc.png"
      />
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  test('renders correctly', () => {
    const wrapper = shallow(
      <ImgCatch
        width="100"
        height="100"
        src="xx.png"
        blanksrc="http://pic/blanksrc.png"
      />
    );
    expect(wrapper.prop('src')).toBe('xx.png');
  });
});
