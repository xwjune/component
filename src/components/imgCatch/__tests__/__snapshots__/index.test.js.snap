// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImgCatch renders correctly 1`] = `
<img
  alt="图片"
  height="100"
  src="pic.png"
  width="100"
/>
`;
